From 27de66f032c422bced3ae65b85c9691e493be960 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Thu, 9 Aug 2018 16:31:54 +0800
Subject: [PATCH 4470/5242] MLK-19176 drm/imx: dpu: plane: Correct fetchunit's
 src_x/y for linear pix fmts

commit  63cdd6d021a08324b65fa33169e69a85faf52a13 from
https://source.codeaurora.org/external/imx/linux-imx.git

For framebuffer linear pixel formats, we calculate the buffer address
with an offset to do framebuffer start point x/y cropping.  This way,
the src_x/y seen by fetchunit(including prefetch engine, if used) can
be zero.  The reason why it still works with the existing nonzero
settings is very likely that the fetchunit or prefetch engine may
discard the initial invalid pixels by looking at the correct buffer
address, frame stride, frame width and frame height.  However, it
would be good to correct the mistake in driver.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
(cherry picked from commit 8847a9b7c28f1053fdd456fadba000704f05ad7f)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dpu/dpu-plane.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/imx/dpu/dpu-plane.c b/drivers/gpu/drm/imx/dpu/dpu-plane.c
index eef39ad..7f41afb 100644
--- a/drivers/gpu/drm/imx/dpu/dpu-plane.c
+++ b/drivers/gpu/drm/imx/dpu/dpu-plane.c
@@ -478,8 +478,8 @@ static void dpu_plane_atomic_update(struct drm_plane *plane,
 
 	src_w = state->src_w >> 16;
 	src_h = state->src_h >> 16;
-	src_x = state->src_x >> 16;
-	src_y = state->src_y >> 16;
+	src_x = fb->modifier ? (state->src_x >> 16) : 0;
+	src_y = fb->modifier ? (state->src_y >> 16) : 0;
 
 	if (fetchunit_is_fetchdecode(fu)) {
 		if (fetchdecode_need_fetcheco(fu, fb->format->format)) {
-- 
1.7.9.5

